src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -71,7 +71,20 @@ public interface UpcallStubFactory {
         MemorySegment makeStub(MethodHandle target, Arena arena);
     }
 
-    private record LinkRequest(FunctionDescriptor descriptor, LinkerOptions options) {}
+    private record LinkRequest(FunctionDescriptor descriptor, LinkerOptions options) {
+        // Overrides for boot performance
+        @Override
+        public boolean equals(Object obj) {
+            return obj instanceof LinkRequest other &&
```

Comment:
Might be worth checking `obj == this || obj instanceof ...`

Reply:
No. We are always using `new LinkRequest` so this "fast path" will 100% miss.